src/lib/HasuraGraphiQL.tsx

```diff
@@ -0,0 +1,563 @@
+  const [headersCollapsed, setHeadersCollapsed] = React.useState(false);
+  const [relay, setRelay] = React.useState(false);
+  const [codeExporterVisible, setCodeExporterVisible] = React.useState(false);
+  const [explorerVisible, setExplorerVisible] = React.useState(true);
+  const [errorShown, setErrorShown] = React.useState(false);
+  const [error, setError] = React.useState(null);
+  const [responseTime, setResponseTime] = React.useState<number | null>(null);
+  const [responseSize, setResponseSize] = React.useState(0);
+  const [explorerWidth, setExplorerWidth] = React.useState(300);
+  const [resizing, setResizing] = React.useState(false);
+  const [isCached, setIsCached] = React.useState(false);
+
+  const updateHeaders = () => {
+    if (headersInput !== headers) {
+      setSchema(undefined);
+      setLoading(true);
+      setHeaders(headersInput);
+    }
+  };
+
+  const graphQLFetcher = makeFetcher(
+    url,
+    url.replace("http", "ws"),
+    transformHeaders(headers),
+    async function customFetch(...args) {
+      setIsCached(hasCacheDirective(query));
+      setResponseTime(null);
+      let start = Date.now();
+      let returnedPromise = fetch(...args);
+      let res = await returnedPromise;
+      setResponseTime(Date.now() - start);
+      let cloned = res.clone();
+      setResponseSize(JSON.stringify(await cloned.json()).length * 2);
+      return returnedPromise;
+    },
+    () => {
+      setResponseTime(null);
+      return null;
+    }
+  );
+
+  function ErrorNotification() {
+    return (
+      <div className="hasura-graphiql-notifications-wrapper">
+        <div className="hasura-graphiql-notification-tr">
+          <div className="hasura-graphiql-notification-inner">
+            <h4 className="hasura-graphiql-notification-title">
+              Schema Introspection Error
+            </h4>
+            <div className="hasura-graphiql-notification-message">
+              {error ? error : "Error loading schema"}
+            </div>
+            <span
+              className="hasura-graphiql-notification-dismiss"
+              onClick={() => setErrorShown(false)}
+            >
+              ×
+            </span>
+          </div>
+        </div>
+      </div>
+    );
+  }
+
+  const extraButtons = () => {
+    const buttons = [
+      {
+        label: "Explorer",
+        title: "Toggle Explorer",
+        onClick: () => setExplorerVisible(!explorerVisible),
+      },
+      {
+        label: "Code Exporter",
+        title: "Toggle Code Exporter",
+        onClick: () => setCodeExporterVisible(!codeExporterVisible),
+      },
+    ];
+    if (isCloud)
+      buttons.push({
+        label: "Cache",
+        title: "Cache the response of this query",
+        onClick: () => setQuery(toggleCacheDirective(query)),
+      });
+    return buttons.map((b) => {
+      return <GraphiQL.Button key={b.label} {...b} />;
+    });
+  };
+
+  React.useEffect(() => {
+    if (loading)
```

Comment:
Great catch! This is a remnant from the initial version which worked differently and does not make sense now. I can safely remove the `if(loading)` and the `loading` from dependencies. Can also remove `url` from dependencies since it's read-only now. Also, will make `url` a prop.